Special case of exact approximation in check functions

jbriales opened the pull request:
Added some comments and control for the special case
when the Taylor expansion is exact.
This may arise when using linear manifolds, such as `tangentspacefactory` or `normalspacefactory`
More comments in the Manopt thread:
https://groups.google.com/forum/#!topic/manopttoolbox/ls2EmvkaHjQ

NicolasBoumal commented on manopt/tools/checkhessian.m:
```diff
@@ -100,18 +100,31 @@ function checkhessian(problem, x, d)
+      range = 1:numel(h);
+      poly = polyfit(log10(h), err, 1);
+      poly(end) = log10(poly(end)); % Set mean error in log scale for plot
+      warning('The model seems exact! Error is zero (near numerical precision) everywhere')
```
Since nothing wrong is happening, I would recommend not using a warning. Instead, please use:

fprintf('The quadratic model appears to be exact (within numerical precision), so that the slope computation is irrelevant.\n')

NicolasBoumal commented on manopt/tools/checkdiff.m:
```diff
@@ -114,5 +114,7 @@ function checkdiff(problem, x, d, force_gradient)
     fprintf('The slope should be 2. It appears to be: %g.\n', poly(1));
     fprintf(['If it is far from 2, then directional derivatives ' ...
              'might be erroneous.\n']);
+    fprintf(['If it is very close to 0, then the modeled funtion is '...
```
Since this scenario is quite rare, could you please place this fprintf(..) inside and if .. end statement, so that it is displayed only if there is a chance that it will be relevant for the user?

Also: is it possible that the slope would turn out to be NaN instead of 0 in the targeted case?

jbriales replied:
Right!
Actually a linear model should be quite unusual (probably even more unusual than the quadratic one), and would be unbounded anyway, so the practical interest is quite low.
However, just for completeness, I think it's fine to have this special case considered as well (it also provides a more correct analysis considering the numerical approximation framework deployed for the test). This complements the theoretical exposition in the Manopt documentation and might came in handy for those just learning the mechanics.

Then, regarding the NaN you're right again. I've added an if-else structure, as in the case of the checkhessian, which simply checks if the numerical errors are low enough to point towards a potential exact model. I think 1e-12 should be low enough to not to produce false positives and still having a good margin for detection of exact models (where actual errors are close to eps, 1e-15 to 1e-16).

NicolasBoumal commented on manopt/tools/checkhessian.m:
```diff
@@ -100,18 +100,31 @@ function checkhessian(problem, x, d)
     hold all;
-    loglog(h(range), 10.^polyval(poly, log10(h(range))), 'LineWidth', 3);
+        loglog(h(range), 10.^polyval(poly, log10(h(range))), ...
+               'r-', 'LineWidth', 3);
```
If the 'r-' option is not passed, thanks to the 'hold all' commande, Matlab will use the next default color, which looks a lot nicer than bright red in modern versions of Matlab.

jbriales replied:
That's definitely right! I suppose maybe I copied it from a previous version, I'll drop it.

NicolasBoumal commented on manopt/tools/checkhessian.m:
```diff
@@ -100,18 +100,31 @@ function checkhessian(problem, x, d)
     fprintf('The slope should be 3. It appears to be: %g.\n', poly(1));
     fprintf(['If it is far from 3, then directional derivatives or ' ...
              'the Hessian might be erroneous.\n']);
+    fprintf(['If it is very close to 0, then the modeled funtion is '...
```
Same as for checkdiff: please place inside an if .. end.

jbriales commented:
Hi Nicolas,
I've fixed some issues using the comments you gave.

In order to give a consistent feedback to the user,
I thought it would be clearer to display either
- the usual message, specifying the slope to expect
- or the message declaring the model was exact.
Also, in case the model is exact this is displayed
in the title of the slope-check figure.
* A warning here: Because of the way the title is built,
  this doesn't apply to checkgradient, but I didn't want
  to modify anything further in the current code :)

I think the feedback should be nice a clear now
for the rare exact cases whereas the code remains simple.

Just for quick testing of where these changes apply,
I added two very simple tests which serve
to show these special cases (the most straightforward example
was to simply use the Euclidean factory, so that's what I did):
- `Test_linearFunction`
- `Test_quadraticFunction`

Let me know if it's fine like this, or any detail should be fixed.

NicolasBoumal commented on manopt/tools/checkdiff.m:
```diff
@@ -101,18 +101,39 @@ function checkdiff(problem, x, d, force_gradient)
     loglog(h(range), 10.^polyval(poly, log10(h(range))), 'LineWidth', 3);
     hold off;
     
+    if ~exist('isModelExact','var') || ~isModelExact
```
In what scenario would it happen that the variable isModelExact does not exist? If that is impossible (as I suspect), please simplify the if condition.

jbriales replied:
Hi,
effectively it should always exist now, I added it just for robustness since the default behavior so far has been that this variable didn't exist. Actually I wasn't completely sure if you would agree with this "conditional" arrangement of the `fprintf`, so this was somehow testing.
Now, since it's really not necessary if you agree with everything else I will drop this redundat condition :)

NicolasBoumal commented on manopt/tools/checkhessian.m:
```diff
@@ -100,15 +100,33 @@ function checkhessian(problem, x, d)
+        loglog(h(range), 10.^polyval(poly, log10(h(range))), 'LineWidth', 3);
     hold off;
     
+    if ~exist('isModelExact','var') || ~isModelExact
```
Same question.

NicolasBoumal commented:
Thanks!
